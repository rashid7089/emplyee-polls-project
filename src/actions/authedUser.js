// types of actions
export const SET_AUTHED_USER = "SET_AUTHED_USER";

// action creators
export function setAuthedUser(id) {
    return {
        type: SET_AUTHED_USER,
        id,
    };
}