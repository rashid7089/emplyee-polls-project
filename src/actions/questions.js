import { _getQuestions } from "../utils/_DATA";

// types of actions
export const GET_QUESTIONS = "GET_QUESTIONS";

// action creators
export function getQuestions(questions) {
    return {
        type: GET_QUESTIONS,
        questions,
    };
}

// thunks action creators
export const handleInitialData_Questions = () => (dispatch) => {
    return _getQuestions().then((questions) => {
        dispatch(getQuestions(questions));
    });
}

