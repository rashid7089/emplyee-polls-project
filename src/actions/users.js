import { _getUsers } from "../utils/_DATA";

// types of actions
export const GET_USERS = "GET_USERS";
export const ADD_USER_QUESTION = "ADD_USER_QUESTION";
export const SET_USER_ANSWER = "SET_USER_ANSWER";
export const REMOVE_USER_ANSWER = "REMOVE_USER_ANSWER";

// action creators
export function getUsers(users) {
    return {
        type: GET_USERS,
        users,
    };
}

export function setUserAnswer(answer, question_id, authedUser) {
    return {
        type: SET_USER_ANSWER,
        answer,
        question_id,
        authedUser,
    };
}

export function removeUserAnswer(question_id, authedUser) {
    return {
        type: REMOVE_USER_ANSWER,
        question_id,
        authedUser,
    };
}


// thunks action creators
export const handleInitialData_Users = () => (dispatch) => {
    return _getUsers().then((users) => {
        dispatch(getUsers(users));
    });
}

export function handleSetUserAnswer(question_id, answer) {
    return (dispatch, getState) => {
        const { authedUser } = getState();
        if (answer !== "") dispatch(setUserAnswer(answer, question_id, authedUser));
        else dispatch(removeUserAnswer(question_id, authedUser));
    };
}


