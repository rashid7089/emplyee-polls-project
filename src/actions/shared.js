import { handleInitialData_Questions } from "./questions";
import { handleInitialData_Users } from "./users";
import { showLoading, hideLoading } from 'react-redux-loading-bar'
import { _saveQuestionAnswer, _saveQuestion } from "../utils/_DATA";

// action types
export const ADD_VOTE = "ADD_VOTE";
export const REMOVE_VOTE = "REMOVE_VOTE";
export const CREATE_NEW_QUESTION = "CREATE_NEW_QUESTION";

// action creater
export const createNewQuestion = (question) => {
    return {
        type: CREATE_NEW_QUESTION,
        question,
    };
}

export function addQuestionVote(authedUser, question_id, answer) {
    return {
        type: ADD_VOTE,
        authedUser,
        question_id,
        answer,
    };
}

export function removeQuestionVote(authedUser, question_id) {
    return {
        type: REMOVE_VOTE,
        authedUser,
        question_id,
        answer:""
    };
}

// thunks action creators
export function handleInitialData() {
    return (dispatch) => {
        dispatch(showLoading());
        return Promise.all([dispatch(handleInitialData_Users()), dispatch(handleInitialData_Questions())])
        .catch((e) => {
            console.log("Error loading initial data", e);
        }).finally(() => dispatch(hideLoading()));
    }
}

export function handleSaveQuestionAnswer(authedUser, question_id, answer) {
    return (dispatch) => {
        dispatch(addQuestionVote(authedUser, question_id, answer));
        return _saveQuestionAnswer({ authedUser, question_id, answer }).catch((e) => {
            alert("Error saving question answer, please try again");
            dispatch(removeQuestionVote(authedUser, question_id));
        });
    }
}

export const handleCreateNewQuestion = (authedUser, optionOneText, optionTwoText) => (dispatch) => {
    dispatch(showLoading());
    return _saveQuestion({
        optionOneText,
        optionTwoText,
        author: authedUser,
    }).then((question) => {
        dispatch(createNewQuestion(question));
    }).catch((e) => {
        console.log("----------- error creating new question", e);
        alert(e);
    }
    ).finally(() => {
        dispatch(hideLoading());
    });
}