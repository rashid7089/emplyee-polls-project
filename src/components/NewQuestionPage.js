import { Typography, Container, TextField, Button } from "@mui/material";
import { useState } from "react";
import { connect } from "react-redux";
import { handleCreateNewQuestion } from "../actions/shared";
import { Navigate, useNavigate } from "react-router-dom";

function NewQuestionPage(props) {
    const { authedUser } = props;
    const [firstOption, setfirstOption] = useState("");
    const [secondOption, setsecondOption] = useState("");
    const [isSubmitted, setisSubmitted] = useState(false); // for testing purposes only

    const navigete = useNavigate();
    
    const UploadQuestion = (e) => {
        e.preventDefault();
        props.dispatch(handleCreateNewQuestion(authedUser, firstOption, secondOption));
        setfirstOption("");
        setsecondOption("");
        setisSubmitted(true); // for testing purposes only
        navigete("/");
    }

    if (!authedUser) <Navigate to="/403" />
    else return ( 
        <Container sx={{textAlign:"center"}} data-testid="newQuestionPage_container">
            <Typography variant="h3" gutterBottom sx={{marginTop:5}}> Would You Rather </Typography>
            <Typography variant="h5" gutterBottom sx={{marginTop:5, marginBottom:5, color:"gray"}}> Create Your Own Poll </Typography>

            <form onSubmit={UploadQuestion} data-testid="form">

            {/* for testing purposes only ---------------------------------- */}
                {isSubmitted && <div data-testid="submittedSign" />} 
                <input // for testing purposes only
                    type="hidden"
                    value={firstOption}
                    data-testid="firstOption"
                    onChange={(e) => setfirstOption(e.target.value)}
                />
                <input // for testing purposes only
                    type="hidden"
                    value={secondOption}
                    data-testid="secondOption"
                    onChange={(e) => setsecondOption(e.target.value)}
                />
            {/* ------------------------------------------------------------ */}

                <TextField 
                    id="outlined-basic" 
                    label="First Option" 
                    variant="outlined" 
                    fullWidth 
                    sx={{marginBottom:2}}
                    value={firstOption} 
                    onChange={(e) => setfirstOption(e.target.value)}
                 />
                <TextField 
                    id="outlined-basic" 
                    label="Second Option" 
                    variant="outlined" 
                    fullWidth 
                    sx={{marginBottom:2}}
                    value={secondOption}
                    onChange={(e) => setsecondOption(e.target.value)}
                 />
                <Button disabled={!firstOption && !secondOption} type="submit" variant="contained" fullWidth sx={{marginBottom:2}}> Upload The Question </Button>
            </form>
           
        </Container>
     );
}

const mapStateToProps = ({authedUser}) => {
    return {
        authedUser
    }
}

export default connect(mapStateToProps)(NewQuestionPage);