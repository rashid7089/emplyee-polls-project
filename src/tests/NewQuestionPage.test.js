import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../reduxStore';
import NewQuestionPage from '../components/NewQuestionPage';
import { BrowserRouter } from 'react-router-dom';
import { setAuthedUser } from '../actions/authedUser';


describe('NewQuestionPage Element', () => { 
      
    it('NewQuestionPage Element must match the snapshot', () => {
        const utils = render(
            <BrowserRouter>
                <Provider store={store}>
                    <NewQuestionPage />
                </Provider>
            </BrowserRouter>
        );
        expect(utils).toMatchSnapshot();
    })

    it('will run submit function', async() => {
        //#region change alert function to avoid it in the test
            const jsdomAlert = window.alert;
            window.alert = () => {};
        //#endregion
    
        store.dispatch(setAuthedUser("johndoe"));

        const utils = render(
            <Provider store={store}>
            <BrowserRouter>
                    <NewQuestionPage />
                </BrowserRouter>
            </Provider>
        );

        const firstFeild = utils.queryByTestId('firstOption');
        const secondFeild = utils.queryByTestId('secondOption');
        const form = utils.queryByTestId('form');
        
        fireEvent.change(firstFeild, {target: {value: "i hate cats"}});
        fireEvent.change(secondFeild, {target: {value: "i hate dog"}});
        fireEvent.submit(form);
        
        const submittedSign = await screen.findByTestId('submittedSign');

        window.alert = jsdomAlert; // restore alert function
        expect(submittedSign).toBeInTheDocument();
    })

    it('newQuestionPage_container should not appear if user is not sign in', () => {
        //#region change alert function to avoid it in the test
            const jsdomAlert = window.alert;
            window.alert = () => {};
        //#endregion    

        store.dispatch(setAuthedUser(""));
        const utils = render(
            <BrowserRouter>
                <Provider store={store}>
                    <NewQuestionPage />
                </Provider>
            </BrowserRouter>
        );
        const newQuestionPage_container = utils.queryByTestId('newQuestionPage_container');
        
        window.alert = jsdomAlert;  // restore alert function
        expect(newQuestionPage_container).not.toBeInTheDocument();
    })

 })


