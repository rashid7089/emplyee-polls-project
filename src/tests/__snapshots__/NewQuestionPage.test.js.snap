// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewQuestionPage Element NewQuestionPage Element must match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-zkjlib-MuiContainer-root"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1 MuiTypography-gutterBottom css-a4d0r7-MuiTypography-root"
        >
           403 
        </h1>
        <h1
          class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
        >
           You are not authrized to see this page 
        </h1>
        <h1
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1xkye6-MuiTypography-root"
        >
           please sign in first
        </h1>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-zkjlib-MuiContainer-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h1 MuiTypography-gutterBottom css-a4d0r7-MuiTypography-root"
      >
         403 
      </h1>
      <h1
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
      >
         You are not authrized to see this page 
      </h1>
      <h1
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1xkye6-MuiTypography-root"
      >
         please sign in first
      </h1>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
